File: backend/app/services/edit_recipe.py
"""Edit recipe validation and defaults — mirrors frontend EditState ranges."""

from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.catalog import PRESET_NAMES


class EditRecipe(BaseModel):
    preset_index: int = Field(default=0, ge=0, le=15)
    preset_intensity: float = Field(default=100, ge=0, le=100)
    exposure: float = Field(default=0, ge=-2, le=2)
    contrast: float = Field(default=1, ge=0.5, le=1.5)
    saturation: float = Field(default=1, ge=0, le=2)
    temperature: float = Field(default=0, ge=-1, le=1)
    pop: float = Field(default=0, ge=0, le=1)
    sharpness: float = Field(default=1, ge=0, le=2)
    fade: float = Field(default=0, ge=0, le=1)
    vignette: float = Field(default=0, ge=0, le=1)
    grain: float = Field(default=0, ge=0, le=1)
    selective_skin: float = Field(default=0, ge=-1, le=1)
    selective_sky: float = Field(default=0, ge=-1, le=1)
    selective_green: float = Field(default=0, ge=-1, le=1)
    selective_warm: float = Field(default=0, ge=-1, le=1)

    @field_validator("preset_index", mode="before")
    @classmethod
    def clamp_preset(cls, v: Any) -> int:
        try:
            n = int(v)
        except (TypeError, ValueError):
            return 0
        return max(0, min(15, n))

    @classmethod
    def from_loose_dict(cls, data: dict[str, Any]) -> "EditRecipe":
        """Accept camelCase or snake_case keys from LLM output."""
        key_map = {
            "presetIndex": "preset_index",
            "presetIntensity": "preset_intensity",
            "selectiveSkin": "selective_skin",
            "selectiveSky": "selective_sky",
            "selectiveGreen": "selective_green",
            "selectiveWarm": "selective_warm",
        }
        normalized: dict[str, Any] = {}
        for k, v in data.items():
            nk = key_map.get(k, k)
            normalized[nk] = v
        return cls.model_validate(normalized)


def preset_label(index: int) -> str:
    idx = max(0, min(15, index))
    return PRESET_NAMES[idx]

File: frontend/src/lib/editRecipe.ts
import type { EditState } from '../engine/editState';

/** API snake_case edit recipe from /api/v1/suggest-styles */
export type ApiEditRecipe = {
  preset_index: number;
  preset_intensity: number;
  exposure: number;
  contrast: number;
  saturation: number;
  temperature: number;
  pop: number;
  sharpness: number;
  fade: number;
  vignette: number;
  grain: number;
  selective_skin: number;
  selective_sky: number;
  selective_green: number;
  selective_warm: number;
};

export type StyleDirection = {
  id: string;
  label: string;
  tagline: string;
  coach_tip?: string;
  edit: ApiEditRecipe;
};

export type SuggestStylesResponse = {
  directions: StyleDirection[];
  reasoning_tr: string;
  source: 'gemini' | 'fallback' | string;
};

export function apiEditToEditState(edit: ApiEditRecipe): EditState {
  return {
    presetIndex: edit.preset_index,
    presetIntensity: edit.preset_intensity,
    exposure: edit.exposure,
    contrast: edit.contrast,
    saturation: edit.saturation,
    temperature: edit.temperature,
    pop: edit.pop,
    sharpness: edit.sharpness,
    fade: edit.fade,
    vignette: edit.vignette,
    grain: edit.grain,
    selectiveSkin: edit.selective_skin,
    selectiveSky: edit.selective_sky,
    selectiveGreen: edit.selective_green,
    selectiveWarm: edit.selective_warm,
  };
}

/** AI önerisini baseline ile karıştırır; intensity 0 = baseline, 100 = tam öneri. */
export function blendEditState(base: EditState, target: EditState, intensityPct: number): EditState {
  const t = Math.max(0, Math.min(100, intensityPct)) / 100;
  if (t <= 0) return { ...base };
  const lerp = (a: number, b: number) => a + (b - a) * t;
  const samePreset = base.presetIndex === target.presetIndex;
  const presetIntensity = samePreset
    ? lerp(base.presetIntensity, target.presetIntensity)
    : lerp(0, target.presetIntensity);
  return {
    presetIndex: target.presetIndex,
    presetIntensity,
    exposure: lerp(base.exposure, target.exposure),
    contrast: lerp(base.contrast, target.contrast),
    saturation: lerp(base.saturation, target.saturation),
    temperature: lerp(base.temperature, target.temperature),
    pop: lerp(base.pop, target.pop),
    sharpness: lerp(base.sharpness, target.sharpness),
    fade: lerp(base.fade, target.fade),
    vignette: lerp(base.vignette, target.vignette),
    grain: lerp(base.grain, target.grain),
    selectiveSkin: lerp(base.selectiveSkin, target.selectiveSkin),
    selectiveSky: lerp(base.selectiveSky, target.selectiveSky),
    selectiveGreen: lerp(base.selectiveGreen, target.selectiveGreen),
    selectiveWarm: lerp(base.selectiveWarm, target.selectiveWarm),
  };
}
